pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,910 @@
+		/// * SPECIAL CASE: AMM pool does not exist - claim may happen if liq. pool is still active, LP
+		/// shares will not be transferred.
+		///
+		/// User's unclaimable rewards will be transferred back to global pool's account.
+		///
+		/// Parameters:
+		/// - `origin`: account owner of deposit(nft).
+		/// - `deposit_id`: nft id representing deposit in the liq. pool.
+		/// - `yield_farm_id`: yield farm identifier to dithdraw shares from.
+		/// - `asset_pair`: asset pair identifying liq. pool in farm.
+		///
+		/// Emits:
+		/// * `RewardClaimed` if claim happen
+		/// * `SharesWithdrawn` event when successful
+		#[pallet::weight(<T as Config>::WeightInfo::withdraw_shares())]
+		#[transactional]
+		pub fn withdraw_shares(
+			origin: OriginFor<T>,
+			deposit_id: primitives::InstanceId,
+			yield_farm_id: PoolId,
+			asset_pair: AssetPair,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let nft_owner =
+				T::NFTHandler::owner(&T::NftClassId::get(), &deposit_id).ok_or(Error::<T>::CantFindDepositOwner)?;
+
+			ensure!(nft_owner == who, Error::<T>::NotDepositOwner);
+
+			let global_farm_id = T::LiquidityMiningHandler::get_global_farm_id(deposit_id, yield_farm_id)
+				.ok_or(Error::<T>::DepositDataNotFound)?;
+
+			let amm_pool_id = T::AMM::get_pair_id(asset_pair);
+
+			let mut unclaimable_rewards: Balance = 0;
+			if T::LiquidityMiningHandler::is_yield_farm_claimable(global_farm_id, yield_farm_id, amm_pool_id.clone()) {
+				//This should not fail on double claim, we need unclaimable_rewards
+				let fail_on_double_claim = false;
+				let (farm_id, reward_currency, claimed, unclaimable) = T::LiquidityMiningHandler::claim_rewards(
+					who.clone(),
+					deposit_id,
+					yield_farm_id,
+					fail_on_double_claim,
+				)?;
+
+				if !claimed.is_zero() {
+					Self::deposit_event(Event::RewardClaimed {
+						farm_id,
+						yield_farm_id,
+						who: who.clone(),
+						claimed,
+						reward_currency,
+					});
+				}
+
+				unclaimable_rewards = unclaimable;
+			}
+
+			let (farm_id, withdrawn_amount, is_destroyed) =
+				T::LiquidityMiningHandler::withdraw_lp_shares(deposit_id, yield_farm_id, unclaimable_rewards)?;
+
+			let lp_token = Self::get_lp_token(&amm_pool_id)?;
+			if !withdrawn_amount.is_zero() {
+				Self::deposit_event(Event::SharesWithdrawn {
+					farm_id,
+					yield_farm_id,
+					who: who.clone(),
+					lp_token,
+					amount: withdrawn_amount,
+				});
+			}
+
+			if is_destroyed {
+				Self::unlock_lp_tokens(amm_pool_id, who.clone(), withdrawn_amount)?;
+				T::NFTHandler::burn_from(&T::NftClassId::get(), &deposit_id)?;
+
+				Self::deposit_event(Event::DepositDestroyed {
+					who,
+					nft_instance_id: deposit_id,
+				});
+			}
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn account_id_for_all_lp_shares() -> AccountIdOf<T> {
+		<T as pallet::Config>::PalletId::get().into_account()
+	}
+
+	fn get_lp_token(amm_pool_id: &AccountIdOf<T>) -> Result<AssetId, Error<T>> {
```

Comment:
I don't think this fn is necessary. You have `AssetPair` as part of api calls, conver it to `amm_pool_id` and back is unnecessary.

Reply:
good one, it can be indeed simplified.